internal/blockchain/cardano/cardano.go

```diff
@@ -0,0 +1,653 @@
+		return nil, i18n.NewError(ctx, coremsgs.MsgContractLocationInvalid, "'address' not set")
+	}
+	return &cardanoLocation, nil
+}
+
+func (c *Cardano) encodeContractLocation(_ context.Context, location *Location) (result *fftypes.JSONAny, err error) {
+	normalized, err := json.Marshal(location)
+	if err == nil {
+		result = fftypes.JSONAnyPtrBytes(normalized)
+	}
+	return result, err
+}
+
+func (c *Cardano) AddContractListener(ctx context.Context, listener *core.ContractListener, lastProtocolID string) (err error) {
+	subName := fmt.Sprintf("ff-sub-%s-%s", listener.Namespace, listener.ID)
+	firstEvent := string(core.SubOptsFirstEventNewest)
+	if listener.Options != nil {
+		firstEvent = listener.Options.FirstEvent
+	}
+
+	filters := make([]filter, len(listener.Filters))
+	for _, f := range listener.Filters {
+		location, err := c.parseContractLocation(ctx, f.Location)
+		if err != nil {
+			return err
+		}
+		filters = append(filters, filter{
+			eventfilter{
+				Contract:  location.Address,
+				EventPath: f.Event.Name,
+			},
+		})
+	}
+
+	result, err := c.streams.createListener(ctx, c.streamID, subName, firstEvent, filters)
+	listener.BackendID = result.ID
+	return err
+}
+
+func (c *Cardano) DeleteContractListener(ctx context.Context, subscription *core.ContractListener, okNotFound bool) error {
+	return c.streams.deleteListener(ctx, c.streamID, subscription.BackendID)
+}
+
+func (c *Cardano) GetContractListenerStatus(ctx context.Context, namespace, subID string, okNotFound bool) (found bool, detail interface{}, status core.ContractListenerStatus, err error) {
+	l, err := c.streams.getListener(ctx, c.streamID, subID, okNotFound)
+	if err != nil || l == nil {
+		return false, nil, core.ContractListenerStatusUnknown, err
+	}
+	return true, nil, core.ContractListenerStatusUnknown, nil
+}
+
+func (c *Cardano) GetFFIParamValidator(ctx context.Context) (fftypes.FFIParamValidator, error) {
+	// Cardanoconnect does not require any additional validation beyond "JSON Schema correctness" at this time
+	return nil, nil
+}
+
+func (c *Cardano) GenerateEventSignature(ctx context.Context, event *fftypes.FFIEventDefinition) (string, error) {
+	params := []string{}
+	for _, param := range event.Params {
+		params = append(params, param.Schema.JSONObject().GetString("type"))
+	}
+	return fmt.Sprintf("%s(%s)", event.Name, strings.Join(params, ",")), nil
+}
+
+func (c *Cardano) GenerateEventSignatureWithLocation(ctx context.Context, event *fftypes.FFIEventDefinition, location *fftypes.JSONAny) (string, error) {
+	eventSignature, _ := c.GenerateEventSignature(ctx, event)
+
+	if location == nil {
+		return fmt.Sprintf("*:%s", eventSignature), nil
+	}
+
+	parsed, err := c.parseContractLocation(ctx, location)
+	if err != nil {
+		return "", err
+	}
+	return fmt.Sprintf("%s:%s", parsed.Address, eventSignature), nil
+}
+
+func (c *Cardano) GenerateErrorSignature(ctx context.Context, event *fftypes.FFIErrorDefinition) string {
+	// TODO: impl
+	return ""
+}
```

Comment:
Would be interested to understand more how errors are reported in Cardano from smart contract

Reply:
The cardano smart contract doesn't report errors right now, only events. I don't understand the difference between errors and events, and I couldn't find any code in other plugins which streamed errors from the connector.

I've copied the "event signature" code into this function, but that's just a guess at what it's supposed to do. I'd want to see how consumers actually use events to be more confident about that.